envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -126,10 +134,27 @@ class RBACFilterFactory(
         return Permission.newBuilder().setHeader(methodMatch).build()
     }
 
-    private fun mapClientToPrincipal(client: String): Principal {
-        val clientMatch = HeaderMatcher.newBuilder()
-                .setName(incomingPermissionsProperties.clientIdentityHeader).setExactMatch(client).build()
-        return Principal.newBuilder().setHeader(clientMatch).build()
+    private fun mapClientToPrincipals(client: String, snapshot: GlobalSnapshot): List<Principal> {
+        if (client !in incomingPermissionsProperties.sourceIpAuthentication.enabledForServices) {
+            val clientMatch = HeaderMatcher.newBuilder()
+                    .setName(incomingPermissionsProperties.clientIdentityHeader).setExactMatch(client).build()
+
+            return listOf(Principal.newBuilder().setHeader(clientMatch).build())
+        }
+
+        val clientEndpoints = snapshot.endpoints.resources().filterKeys { client == it }.values
```

Comment:
We update all groups when any change occurs, right?
In case we will want to be smart and don't update all groups on every change we have to remember that group have to get updated not only when on changes to upstream but also to downstream.

Reply:
> We update all groups when any change occurs, right?

This is triggered by changes in the Flux with all services or new group. We have to regenerate this because the ip addresses of an endpoint might change so we need to set principals accordingly. We would have to cache the results and get a diff of services that changed to not regenerate everything.

Reply:
That's what I've written.